### PR TITLE
feat(agent): add command explain-the-code to the vscode extension

### PR DESCRIPTION
**Pull request by wwayne**
### In command palette
commanpanel

### In right-click
rightclick

### Open the browser
browser

fix TAB-572

### CODE REVIEW

**Review by wsxiaoys — COMMENTED** [review]
Please gate the feature with settings

**Comment by wwayne** [comment]
Added to the settings, default is false
settings

File: clients/vscode/package.json [hunk 1]
```diff
@@ -69,6 +69,10 @@
       {
         "command": "tabby.notifications.resetMuted",
         "title": "Tabby: Reset notifications marked as \"Don't Show Again\""
+      },
+      {
+        "command": "tabby.ai.explainCodeBlock",
```
Comment (icycodes) [inline clients/vscode/package.json:74]: I think it's better to use `tabby.chat` as prefix: `tabby.chat.explainCodeBlock`

**Review by icycodes — APPROVED** [review]
LGTM

File: clients/vscode/package.json (hunk 1 above)
Comment (wsxiaoys) [reply clients/vscode/package.json:74]: `tabby.experimental.chat.explainCodeBlock`

**Comment by wwayne** [comment]
* when clause doesn't support nested configuration, so I add a new context variable `isExplainCodeEnabled`, it will be updated when status ready & config changed
* adjust the configuration structure in package.json from {} to [{}], only for make sure the `tabby.experimental` is in the bottom

File: clients/vscode/src/TabbyStatusBarItem.ts [hunk 2]
```diff
@@ -148,6 +148,15 @@ export class TabbyStatusBarItem {
...
       this.logger.info("Tabby agent statusChanged", { event });
       this.fsmService.send(event.status);
+      if (event.status === "ready") {
+        const healthState = agent().getServerHealthState();
+        const isChatEnabled = Boolean(healthState?.chat_model);
+        commands.executeCommand("setContext", "chatModeEnabled", isChatEnabled);
+        const configuration = workspace.getConfiguration("tabby");
+        const experimental = configuration.get<Record<string, any>>("experimental", {});
+        const isExplainCodeEnabled = experimental["chat.explainCodeBlock"] || false;
+        commands.executeCommand("setContext", "explainCodeSettingEnabled", isExplainCodeEnabled);
```
Comment (wwayne) [inline clients/vscode/src/TabbyStatusBarItem.ts:158]: new update in 4.17

File: clients/vscode/src/agent.ts [hunk 3]
```diff
@@ -100,6 +100,11 @@ export async function createAgentInstance(context: ExtensionContext): Promise<Ta
         const keybindings = configuration.get<string>("keybindings", "vscode-style");
         agent.updateClientProperties("user", "vscode.keybindings", keybindings);
       }
+      if (event.affectsConfiguration("tabby.experimental")) {
+        const experimental = configuration.get<Record<string, any>>("experimental", {});
+        const isExplainCodeEnabled = experimental["chat.explainCodeBlock"] || false;
+        commands.executeCommand("setContext", "explainCodeSettingEnabled", isExplainCodeEnabled);
+      }
```
Comment (wwayne) [inline clients/vscode/src/agent.ts:107]: new update in 4.17

File: clients/vscode/package.json [hunk 4]
```diff
@@ -124,45 +139,61 @@
...
+        "properties": {
+          "tabby.experimental": {
+            "default": {},
+            "properties": {
+              "chat.explainCodeBlock": {
+                "type": "boolean",
+                "default": false
+              }
+            }
+          }
```
Comment (wwayne) [inline clients/vscode/package.json:193]: new update in 4.17

**Pull request merged**